packages/block-editor/src/hooks/font-size.js

```diff
@@ -0,0 +1,189 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+import { hasBlockSupport } from '@wordpress/blocks';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { PanelBody } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+import { useSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import {
+	getFontSize,
+	getFontSizeClass,
+	getFontSizeObjectByValue,
+	FontSizePicker,
+} from '../components/font-sizes';
+import InspectorControls from '../components/inspector-controls';
+
+const FONT_SIZE_SUPPORT_KEY = '__experimentalFontSize';
+
+/**
+ * Filters registered block settings, extending attributes to include
+ * `fontSize` and `fontWeight` attributes.
+ *
+ * @param  {Object} settings Original block settings
+ * @return {Object}          Filtered block settings
+ */
+function addAttributes( settings ) {
+	if ( ! hasBlockSupport( settings, FONT_SIZE_SUPPORT_KEY ) ) {
+		return settings;
+	}
+
+	// Allow blocks to specify a default value if needed.
+	if ( ! settings.attributes.fontSize ) {
+		Object.assign( settings.attributes, {
+			fontSize: {
+				type: 'string',
+			},
+		} );
+	}
+
+	return settings;
+}
+
+/**
+ * Override props assigned to save component to inject font size.
+ *
+ * @param  {Object} props      Additional props applied to save element
+ * @param  {Object} blockType  Block type
+ * @param  {Object} attributes Block attributes
+ * @return {Object}            Filtered props applied to save element
+ */
+function addSaveProps( props, blockType, attributes ) {
+	if ( ! hasBlockSupport( blockType, FONT_SIZE_SUPPORT_KEY ) ) {
+		return props;
+	}
+
+	const { fontSize } = attributes;
+
+	const fontSizeClass = getFontSizeClass( fontSize );
+
+	props.className = classnames( props.className, fontSizeClass );
+
+	return props;
+}
+
+/**
+ * Filters registered block settings to expand the block edit wrapper
+ * by applying the desired styles and classnames.
+ *
+ * @param  {Object} settings Original block settings
+ * @return {Object}          Filtered block settings
+ */
+function addEditProps( settings ) {
+	if ( ! hasBlockSupport( settings, FONT_SIZE_SUPPORT_KEY ) ) {
+		return settings;
+	}
+
+	const existingGetEditWrapperProps = settings.getEditWrapperProps;
+	settings.getEditWrapperProps = ( attributes ) => {
+		let props = {};
+		if ( existingGetEditWrapperProps ) {
+			props = existingGetEditWrapperProps( attributes );
+		}
+		return addSaveProps( props, settings, attributes );
+	};
+
+	return settings;
+}
+
+const FontSizeInspectorControl = ( { value, onChange } ) => (
+	<InspectorControls key="inspector-controls">
+		<PanelBody title={ __( 'Text settings' ) }>
```

Comment:
In #20775 we're adding the line height hook into a new "Typography" panel. These two hooks need to be added to the same panel so we need to sync when one of the PR merges.

Reply:
It looks like that's ready to merge. I'd let that get in first and then we can work on this and discuss how to merge things.

I was following the thread and my current thoughts are:

- In terms of how blocks would opt-in =>
  - font settings are one thing: line height, font size, font weight, font family (you don't opt-in to each of them separately).
  - drop cap is a separate support key.
- In terms of UI => 
  - All font settings (font size, weight, line-height, font family) would be enabled at once if you opt-in into font and will use the same panel.
  - The drop cap should also use the same panel but will be enabled separately.

How does this sound?

cc @ItsJonQ

Reply:
I don't have a precise answer on the opt-in flag. Maybe all these flags are part of the same "style" flag but I think at this moment, it's too soon to make the best decision. We should just continue with experimental flags and adjust when we want to stabilize the approach.

In terms of UI, it makes sense to gather all typography settings IMO. Potentially, each hook would export a "component" or something like that and everything is applied on the same "style" hook.

Reply:
> In terms of how blocks would opt-in... In terms of UI ...

@nosolosw That makes sense to me! Attributes like `line-height`, `font-weight`, etc... are very generic. As a user, I would expect them to be available anywhere I could adjust text.

Drop cap on the other hand... is very specific.

Based on that model, I think keeping it separate makes sense!

<3

Reply:
Looks like the other PR is merged, time to rebase and see how we can merge things :)